### ISSUE TITLE
Support for Docker install for Node-RED

### ISSUE THREAD

**Issue by vielmetti** [node-red/node-red#603]
It would be mighty nice to have Node-RED build and run neatly under Docker, and to have the documentation and/or scripts for that included in the distribution (somewhere). There has been discussion on the mailing list.

A few bits that are a useful start:

* A Dockerfile from @dceejay at https://gist.github.com/dceejay/9435867 that uses node:0.10.36-slim
* A Dockerfile from @crouchingbadger at https://gist.github.com/crouchingbadger/c0a0e3eba4242baebd07 that uses node:0.10.38
* A blog post from http://ivyco.blogspot.com/2015/03/docker-awesomeness.html that describes Docker on a Raspberry Pi running Node-RED

Opening up an issue here to deal with that...if it belongs in another repo we can move it.

**Comment by vielmetti** [comment]
Also a Dockerfile from @bortels from mid last summer using Ubuntu 12.04 and Node-RED 0.8.1.

https://github.com/bortels/docker-node-red/blob/master/Dockerfile

**Comment by bortels** [comment]
IIRC, the biggest issue (unsolved) was the web sockets. This was quite some time ago; I should go look at it again, it may have simply been docker being so young... Or just me not quite getting it right.

**Comment by vielmetti** [comment]
And of course now that 0.10.6 is out, it would be worthwhile to update the Dockerfiles to use that.

**Comment by knolleary** [comment]
This is certainly on our list to do. What has stopped us has (aside from it not being a top of the list priority) is deciding what the image should look like and how it would be used. Talking it through with @dceejay today, I think we came up with at least 3 different patterns of usage - all valid, but all very different in what it would expect of a node-red docker image.

Short for time right now, but will endeavour to list out the scenarios this evening so we can discuss what approach we want to take.

**Comment by jmls** [comment]
FYI - we have node-red running under a docker instance, but as part of an embedded app using strongloop. The websockets seem to work just fine.

However, having said that, one of the major goals for us is to make both nodes and flows loadable from a database, so that we are not constrained by having to have "data" in a docker container - if a db is used as the source / target then the nodes & flows can be updated / changed / added without having to do a rebuild for the container. (and no, I don't want to have to -v them ;)  - we want the docker image to be pulled onto any machine, connect to the database and go, not be beholden to having to have certain files in certain places on the filesystem

thanks

**Comment by dceejay** [comment]
@vielmetti - "my" gist pulls from latest stable - so if re-built now it should be v0.10.6...
@jmls  - yes all part of the chat @knolleary and I were having - about how easy/complex we need to make it.

**Comment by dceejay** [comment]
Some current thoughts for comment

### Simple
add an on-build style docker file to root of main project

**Advantages**

 - easy to build
 - it can pick up local already installed customisations
 - quick way to "clone" a working system
 - works best if Node-RED installed without -g into a simple node-red dir as can then copy in any extra nodes already installed locally , flow files etc. But will work from /usr/lib/node_modules/node-red - just awkward

**Disadvantages**

  - works best if... (as above...). This is a tiny bit more complex if you want to also get local flow files into the image (as they are now normally in $HOME/.node-red and Dockerfiles can't just accept $HOME as a scriptable variable)
  - not pre-built (but is easy to do :-)

### Just a Dockerfile
just provide a "full" Dockerfile - i.e. not using the on-build method

**Advantages**

  - could document Dockerfile line by line to show options
  - could add extra npms to the file
  - fully customisable
  - show how to add extra commands (eg non- Node-RED) commands
  - would work as a "gist" that could be pointed at by a docker build command

**Disadvantages**

  - more complicated than "simple" if you just want to build something that works, or make a container of your working setup.
  - not pre-built

### Pre-built on Docker Hub
provide a pre-built image on Docker Hub of latest stable version

**Advantages**

  - no need to build - just use Docker run node-red/node-red (sort of thing)
  - Dockerfile is automatically available if someone wants to copy/modify

**Disadvantages**

  - would need sensible defaults for ports and volumes (storage for flow files) to be sorted in advance - see below

### Volumes - storage of flow files
The storage of flows files in this environment from simple but dangerous to complex but safer, with several levels of head scratching in between... indeed to my mind this is the most complex aspect of this ... how to make it easy for beginners but show folks how to extend it.

  - Do nothing - flows exist in the container, when you restart they are wiped... Yikes ! Let's not go there.
  - Simple volume - flows exist outside the container, so you can stop, restart, and even upgrade the container without losing the flows. (I quite like this as a level-0).
 BUT if you recreate the container then the link to the external storage gets broken and while the data still exists on the host drive (so you can re-find them ... eventually) it doesn't stay linked easily. 
  - Named volume - flows exist outside the container in a named location on the host machine. Good for some use cases - but is less transportable. (This approach is normally used for test/debug scenarios...) If you're not careful you may share files between machines you may not intend to... if you use the same shared space.
  - Storage container - by setting up a non-running - but named container you can create a moveable but linkable space that is much
  more in the Docker style of doing tings... however you now have to set up two containers and link them, and the documentation starts getting much more complex...
  - Database container - as previous but using a running database container to store everything - which can then use the database pluggable storage within Node-RED, and could work across networks etc etc... loads of documentation (and indeed testing required to see if it even works :-)

### On-build plus meta package
One quite neat way of combining some of these is to use the simple on-build style  Dockerfile - but combine it with "just" a local meta package.json file that contains links to a list of required npms (including node-red of course). So minimal footprint required to build a working image. - but again does some local customisation and building.

### Automated builds

If we can get a sensible default build sorted then we can let docker hub automatically build images from github

        https://docs.docker.com/docker-hub/builds/

That is if we want to use the github version - not sure how we only use "stable" for that.

**Comment by jmls** [comment]
#3 +1 ;)

storage : have the choice of either a storage container or database container. Using docker compose, this can be handled in a single yaml file, with the data container selected by default, but comments to show how to enable a db container

We would certainly want to use a db container.

we've played an awful lot with docker over the past year and it's getting simpler by the day to manage "complex" environments

**Comment by knolleary** [comment]
I think a key part of this is to think about what users want to be able to do with a dockerized version of node-red - rather than just what shape it may be.
The main use case I would like us to satisfy is:

> A user wants to try node-red with minimum effort

Having a pre-built image is key to this. By default, keeping it all in the container is okay - restarting the container (stop/start) will not wipe your data. Only if you kill the container so it no longer exists, then yes it gets wiped - but that's what you expect when you kill a container. But that's good - clean up after yourself and not worry about bits of data lying around. As an out-of-the-box option, not config needed, we should do this.

If that image also has a means to point it at some external storage, then that's good too - and it ought to be via volume or named storage container in the first instance. Pointing at a db container adds quite a lot of complexity - the base image would need the storage plugins to be included and some means to pick which one you want. We could get there eventually, but I don't consider that a priority.

Experience tells us that there are lots of ways you can run node-red, and lots of ways you may want to run it in the context of docker. If we try creating a single thing that caters for lots of these uses, it risks not doing any one use case well.

I would rather we didn't jump straight to the complex scenarios - lets get the basics done first.

The on-build/meta package approach is a good one - make it easy for users to create their own customised containers based on that.

But... would like to think through how we could make a pre-built image customisable - in particular, enabling adminAuth.

**Comment by dceejay** [comment]
Don't disagree with most of that :-)... But I do think the basic volume for user data would be "a good thing". 1) it does keep the date somewhere in case of " oops I overwrote the container", albeit slightly hard to find, but recoverable. 2) you can upgrade the underlying image and then restart the container... Means people can keep up with our releases without having to save flows out and over to a new machine image somehow. 3) updating the underlying contains means any other images based on it get updated too, rather than having to update each image individually.

**Comment by vielmetti** [comment]
Just to layer another use case on top of all of these. One of my goals eventually is to have Node-RED running inside docker, and then to have those containers provisioned using salt and salt-cloud. The relevant open issue on the other project is https://github.com/saltstack/salt/issues/22322 for reference (there's a bunch of work to be done before this can really work).

One of the things I see about docker as a piece of the Node-RED system is getting some kind of isolation between independent Node-RED development efforts. For example, the old development Mac I ran on had a single Node-RED instance with 8 or 9 separate pages all really doing separate applications all jumbled together into a single image. It would be great for production to peel out a single self-contained page, minimize its dependencies, package up just the essentials of what it's doing, and deploy it to a stable environment.

**Comment by dceejay** [comment]
If it's running inside and nicely isolated then surely it doesn't care what it is running on - so we can leave the Salt side of things to you ;-)
The on-build style mentioned above can pretty much do that... you can get it all running how you want, then clone it into a container... repeat for the other 8 pages you have...

**Comment by xadh00m** [comment]
Great discussion. I´d like to add our planned usage scenario:

* We have a docker host VMWare image where we want to deploy probably a dozen node-red instances which manage a logical piece each in a complex network environment.
* Our flow files are all hosted in individual git repositories for version control and data safety.
* Development: The developer checks out and runs a flow for testing and improvements on his local machine in a staging environment. When he has finished he can push the new flow to the docker host (e.g. with deis/dokku) which replaces the current docker instance related to this repository.

For us it seems to make sense to keep a Dockerfile in the flow file repository which takes care of setting up node-red in a container and copying the flow and settings file. We would use node-red containers only as a service with immutable flow files and therefore without the web frontend.

**Comment by cpswan** [comment]
I've been doing an unofficial Dockerised Node-Red for a little while now at https://registry.hub.docker.com/u/cpswan/node-red/

It's just been updated to 0.10.6, and I've vastly simplified the Dockerfile this time around to use the official Node image and the new npm install mechanism for Node-Red.

**Comment by vielmetti** [comment]
I've been using this Docker image on my Raspberry Pi since yesterday:
    https://registry.hub.docker.com/u/nieleyde/rpi-nodered/
It's from @nieleyde and it seems to work fine, though it doesn't have a Dockerfile that I can find so it's not as useful for learning.

**Comment by vielmetti** [comment]
The source for @nieleyde 's rpi-nodered is at https://github.com/nieleyde/rpi-nodered

The slight complication of rebuilding it is that it's running on Arch Linux for ARM and that distro no longer has the 0.10.x node.js available. https://github.com/nieleyde/rpi-nodered/issues/1 is the relevant issue.

**Comment by muka** [comment]
Hi,
I've been working on this for a while https://github.com/muka/redzilla
Contains a docker image and a build script and a minimal nodejs based api for further customizations (like multi instances)

**Comment by dceejay** [comment]
+1 Nice

**Comment by vielmetti** [comment]
Going back to the question of "volumes for storage of flow files", another part of what I seem to be heading towards is a volume with a set of nodes in it, so that I can extend or experiment with new nodes without building a whole new base configuration. I think that's a small task but I'm not sure yet.

**Comment by dceejay** [comment]
yes - fairly easy, just use the -volume option to mount a directory into the image...
https://docs.docker.com/userguide/dockervolumes/
the simplest way is to just mount a directory from the host (fine if you own the host of course) - though the docs show how to use a data container to do it and then link the containers.

**Comment by vielmetti** [comment]
Thanks @dceejay @jmis for help. I managed to put together a simple Docker Compose file that launches Node-RED and mosquitto both on a Raspberry Pi (so that I have a local MQTT broker) based on images from @nieleyde , you can see that here at

http://vielmetti.typepad.com/vacuum/2015/07/bringing-up-node-red-and-mosquitto-under-docker-compose-on-a-raspberry-pi.html

You'd need a slightly different version picking up different images to run on x86, but this got me going and there is an example.

**Comment by spants** [comment]
You can also try my version = NodeRed v0.12.4 with a number of built in Nodes, support for storing the flows externally and and example of passing a USB serial device through to the docker.

###https://hub.docker.com/r/spants/nodered/

example commandline:
```sh
docker run -d --name nodered -p 1880:1880 --device=/dev/ttyACM0:/dev/ttyACM0  -v /mnt/user/docker/appdata/nodered:/data:rw spants/nodered
```

* serial device ttyACM0 passed through to NodeRed docker
* flow stored in /mnt/user/docker/appdata/nodered on the host (linked to /data in the docker image)

**Comment by jthomas** [comment]
Reviewing the comments, feedback and suggestions above, we've now started to publish official Node-RED Docker images. The existing Docker images the community had already published were a great source of inspiration.

*Images:* https://hub.docker.com/r/nodered/node-red-docker/tags/
*Repository:* https://github.com/node-red/node-red-docker

We've focused on enabling new users to get started quickly, with sensible defaults and customisation via environment variables. There are three different images, a full-Linux base image, a smaller image using Alpine and a Raspberry Pi compatible version.

The project's README has a comprehensive introduction to using and customising the images.

**Feel free to use this issue to provide feedback, feature requests or concerns (...or send a PR 😄).**

Future Plans
---

- **Node-RED Version Images.** We've set up a CI to build the images automatically whenever Node-RED is updated. In the future, we would like to configure this to publish the new image as that version tag. 
- **Node-RED Official Docker Hub Repo.** Follow the process to register the Node-RED organisation as an [official Docker Hub repository](https://docs.docker.com/docker-hub/official_repos/). 
- **Docker Compose Samples.** Publish ```docker-compose.yml``` files for common patterns to help users spin up Node-RED and connected services e.g. MQTT server.
